### PR TITLE
PHPStan (Level 2) on classes/

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | PHPStan (Level 2) on classes/
| Type?             | improvement
| Category?         | CO
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | Relative to #16471
| How to test?      | CI is :green_circle: 

**:notebook: BC Breaks :**
* Added 3 methods in interface `WebserviceOutputInterface`
  * `public function setLanguages($languages);`
  * `public function renderAssociationWrapperHeader();`
  * `public function renderAssociationWrapperFooter();`
* Added 1 argument to two methods in interface `WebserviceOutputInterface`
  * `public function renderNodeHeader($obj, $params, $more_attr = null);` becomes `public function renderNodeHeader($obj, $params, $more_attr = null, $has_child = true);` 
  * `public function renderAssociationHeader($obj, $params, $assoc_name);` becomes `public function renderAssociationHeader($obj, $params, $assoc_name, $closed_tags = false);`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/26164)

### CODE REVIEW

File: classes/Context.php [hunk 1]
```diff
@@ -308,8 +314,8 @@ public static function getContext()
     }
 
     /**
-     * @param $testInstance Context
-     * Unit testing purpose only
+     * @param Context $testInstance
+     *                              Unit testing purpose only
```
Comment (matks) [inline classes/Context.php]: This is mandatory for phpstan?

File: classes/order/Order.php [hunk 2]
```diff
@@ -1242,8 +1243,9 @@ public function addCartRule($id_cart_rule, $name, $values, $id_order_invoice = 0
             $cart_rule = new CartRule($id_cart_rule);
             $free_shipping = $cart_rule->free_shipping;
         }
-        $order_cart_rule->free_shipping = (int) $free_shipping;
-        $order_cart_rule->add();
+        $order_cart_rule->free_shipping = (bool) $free_shipping;
```
Comment (matks) [inline classes/order/Order.php:1246]: Warning: does ObjectModel expects `0` or `false` ?

I know Configuration wants 0 or 1 for example.

File: classes/webservice/WebserviceOutputInterface.php [hunk 3]
```diff
@@ -54,4 +54,10 @@ public function renderErrorsHeader();
     public function renderErrorsFooter();
 
     public function renderErrors($message, $code = null);
+
+    public function setLanguages($languages);
+
+    public function renderAssociationWrapperHeader();
+
+    public function renderAssociationWrapperFooter();
```
Comment (matks) [inline classes/webservice/WebserviceOutputInterface.php:62]: Adding new public functions to an interface can be a pretty big BC break, even if it's actually a bugfix.

**Review by matks — COMMENTED** [review]
Great work ! I have raised 3 possible warnings. Please check them ;) so we know this PR does not introduce unwanted issues

File: classes/Context.php (hunk 1 above)
Comment (Progi1984) [reply classes/Context.php]: No, it's fixed

File: classes/webservice/WebserviceOutputInterface.php (hunk 3 above)
Comment (Progi1984) [reply classes/webservice/WebserviceOutputInterface.php:62]: Affirmative, I document it.

File: classes/order/Order.php (hunk 2 above)
Comment (Progi1984) [reply classes/order/Order.php:1246]: The property is the type `bool` with a boolean validator 
image

You can put an integer (1 or 0) or a boolean. But it's more consistent to set a boolean as boolean.

**Review by PierreRambaud — DISMISSED** [review]

**Comment by atomiix** [comment]
Can you rebase your PR @Progi1984? :)

**Review by PierreRambaud — DISMISSED** [review]

File: classes/Cart.php [hunk 4]
```diff
@@ -2053,7 +2053,10 @@ public static function getTotalCart($id_cart, $use_tax_display = false, $type =
 
         $with_taxes = $use_tax_display ? $cart->_taxCalculationMethod != PS_TAX_EXC : true;
 
-        return Context::getContext()->getCurrentLocale()->formatPrice($cart->getOrderTotal($with_taxes, $type), Currency::getIsoCodeById((int) $cart->id_currency), false);
+        return Context::getContext()->getCurrentLocale()->formatPrice(
```
Comment (jolelievre) [inline classes/Cart.php:2056]: Why did the third argument disappear?
Or rather why was there one before if the interface only has two parameters? 🤔

Comment (Progi1984) [reply classes/Cart.php:2056]: Since 2018, there are only 2 parameters.

File: classes/log/FileLogger.php [hunk 5]
```diff
@@ -59,9 +59,6 @@ public function setFilename($filename)
 
     /**
      * Log the message.
-     *
-     * @param string message
-     * @param level
      */
     public function getFilename()
```
Comment (jolelievre) [inline classes/log/FileLogger.php:72]: What about the return?

File: classes/stock/StockAvailable.php [hunk 6]
```diff
@@ -106,18 +106,16 @@ class StockAvailableCore extends ObjectModel
     ];
 
     /**
-     * For a given {id_product, id_product_attribute and id_shop}, gets the stock available id associated.
-     *
-     * @param int $id_product
-     * @param int $id_product_attribute Optional
-     * @param int $id_shop Optional
-     *
      * @return int
```
Comment (jolelievre) [inline classes/stock/StockAvailable.php]:
```suggestion
     * @return bool
```

File: classes/stock/StockAvailable.php [hunk 7]
```diff
@@ -564,11 +563,11 @@ public static function updateQuantity($id_product, $id_product_attribute, $delta
      *
      * @param int $id_product
      * @param int $id_product_attribute
-     * @param $quantity
+     * @param int $quantity
      * @param int|null $id_shop
      * @param bool $add_movement
      *
-     * @return bool
+     * @return void
```
Comment (jolelievre) [inline classes/stock/StockAvailable.php]:
```suggestion
     * @return false|void
```

**Review by jolelievre — DISMISSED** [review]
Thanks @Progi1984

**Review by PierreRambaud — DISMISSED** [review]

**Comment by Progi1984** [comment]
@boubkerbribri If you have some time for the QA of this PR :)

**Review by PierreRambaud — APPROVED** [review]

**Review by boubkerbribri — APPROVED** [review]
UI tests are all OK.
It's QA :heavy_check_mark:

**Comment by Progi1984** [comment]
Thanks @boubkerbribri

**Pull request merged**